plugins/rapid7_insight_agent/icon_rapid7_insight_agent/util/graphql_api/api_connection.py

```diff
@@ -84,6 +84,92 @@ def unquarantine(self, agent_id: str) -> bool:
+        """
+        Quarantine an agent given a list of agent IDs
+
+        :param agent_hostname_list: List of agent hostnames to quarantine
+        :param advertisement_period: Amount of time, in seconds, to try to take the quarantine action
+
+        :return: Two lists containing asset ids for successful or unsuccessful quarantines
+        """
+
+        # Create empty lists for successful & unsuccessful
+        successful_quarantine = []
+        unsuccessful_quarantine = []
+
+        # Convert the original array of hostnames to IDs
+        agent_id_list = self._convert_hostnames_to_id(agent_hostname_list)
+
+        # For each agent ID in the list, perform quarantine
+        for agent in agent_id_list:
+            quarantine_payload = {
+                "query": "mutation( $orgID:String! $agentID:String! $advPeriod:Long! ) { quarantineAssets( "
+                "orgId:$orgID assetIds: [$agentID] input: {advertisementPeriod: $advPeriod} ) { results { "
+                "assetId failed } } }",
+                "variables": {
+                    "orgID": self.org_key,
+                    "agentID": agent,
+                    "advPeriod": advertisement_period,
+                },
+            }
+
+            results_object = self._post_payload(quarantine_payload)
+            asset_id = results_object.get("data", {}).get("quarantineAssets").get("results")[0].get("assetId")
+
+            if not results_object.get("data").get("quarantineAssets").get("results")[0].get("failed"):
+                successful_quarantine.append(asset_id)
+            else:
+                unsuccessful_quarantine.append(asset_id)
+
+        return successful_quarantine, unsuccessful_quarantine
+
+    def unquarantine_list(self, agent_hostname_list: List[str]) -> Tuple[List[str], List[str]]:
+        """
+        Unquarantine an agent given a list of agent IDs
+        :param agent_hostname_list: List of agent IDs to unquarantine
+
+        :return: Two lists containing asset ids for successful & unsuccessful unquarantine operations.
+        """
+        # Create empty lists for successful & unsuccessful
+        successful_unquarantine = []
+        unsuccessful_unquarantine = []
+
+        # Convert the original array of hostnames to IDs
+        agent_id_list = self._convert_hostnames_to_id(agent_hostname_list)
+
+        # For each agent ID in the list, perform unquarantine
+        for agent_id in agent_id_list:
+            # Check if agent with ID exists
+            payload = {
+                "query": "query( $orgID: String! $agentID: String! ) { assets( orgId: $orgID ids: [$agentID] ){ agent { id quarantineState{ currentState } agentStatus } } }",
+                "variables": {"orgID": self.org_key, "agentID": agent_id},
+            }
+            results_object = self._post_payload(payload)
+            agent = results_object.get("data", {}).get("assets")[0].get("agent")
+            if not agent:
```

Comment:
If this happens part way through the list e.g. for agent 3 out of 5 agents in the list, then it will raise an exception and the user will be unaware that previous agents may have already been successfully quarantined?